src/mlir/driver.cc

```diff
@@ -47,65 +58,71 @@ namespace mlir::verona
     }
   }
 
-  llvm::Error Driver::readAST(const ::ast::Ast& ast)
+  LogicalResult Driver::readAST(const ::ast::Ast& ast)
   {
     auto result = Generator::lower(&context, ast);
     if (!result)
-      return result.takeError();
+    {
+      llvm::logAllUnhandledErrors(result.takeError(), llvm::errs());
+      return failure();
+    }
 
     module = std::move(*result);
 
     if (failed(verify(*module)))
     {
       module->dump();
-      return runtimeError("AST was lowered to invalid MLIR");
+      llvm::errs() << "AST was lowered to invalid MLIR\n";
+      return failure();
     }
 
-    return llvm::Error::success();
+    return success();
   }
 
-  llvm::Error Driver::readMLIR(const std::string& filename)
+  LogicalResult Driver::readMLIR(std::unique_ptr<llvm::MemoryBuffer> buffer)
   {
-    if (filename.empty())
-      return runtimeError("No input filename provided");
-
-    // Read an MLIR file
-    auto srcOrErr = llvm::MemoryBuffer::getFileOrSTDIN(filename);
-
-    if (auto err = srcOrErr.getError())
-      return runtimeError(
-        "Cannot open file " + filename + ": " + err.message());
-
     // Add the input to the source manager and parse it.
     // `parseSourceFile` already includes verification of the IR.
-    sourceManager.AddNewSourceBuffer(std::move(*srcOrErr), llvm::SMLoc());
+    sourceManager.AddNewSourceBuffer(std::move(buffer), llvm::SMLoc());
     module = mlir::parseSourceFile(sourceManager, &context);
+
     if (!module)
-      return runtimeError("Can't load MLIR file");
+    {
+      llvm::errs() << "Can't load MLIR file\n";
+      return failure();
+    }
 
-    return llvm::Error::success();
+    return success();
   }
 
-  llvm::Error Driver::emitMLIR(llvm::StringRef filename)
+  LogicalResult Driver::emitMLIR(llvm::raw_ostream& os)
   {
     assert(module);
 
-    if (filename.empty())
-      return runtimeError("No output filename provided");
-
     if (failed(passManager.run(module.get())))
     {
       module->dump();
-      return runtimeError("Failed to run some passes");
+      llvm::errs() << "Failed to run some passes\n";
+      return failure();
     }
 
-    // Write to the file requested
-    std::error_code error;
-    auto out = llvm::raw_fd_ostream(filename, error);
-    if (error)
-      return runtimeError("Cannot open output filename");
+    module->print(os);
+    return success();
+  }
+
+  LogicalResult Driver::verifyDiagnostics()
```

Comment:
This method is called asynchronously from the user (`verona-mlir`) and so is independent on the kind of errors they emit. It also shouldn't print error messages and should return the errors to the caller to handle.